v1/plugin/plugin.go

```diff
@@ -78,27 +82,140 @@ type StreamCollector interface {
 	GetMetricTypes(Config) ([]Metric, error)
 }
 
-func StartStreamCollector(plugin StreamCollector, name string, version int, opts ...MetaOpt) int {
-	getArgs()
-	opts = append(opts, rpcType(gRPCStream))
-	m := newMeta(collectorType, name, version, opts...)
-	server := grpc.NewServer()
-	proxy := &StreamProxy{
-		plugin:      plugin,
-		pluginProxy: *newPluginProxy(plugin),
+// tlsServerSetup offers functions supporting TLS server setup
+type tlsServerSetup interface {
+	// makeTLSConfig delivers TLS config suitable to use for plugins, excluding
+	// setup of certificates (either subject or root CA certificates).
+	makeTLSConfig() *tls.Config
+	// readRootCAs is a function that delivers root CA certificates for the purpose
+	// of TLS initialization
+	readRootCAs() (*x509.CertPool, error)
+}
+
+// osInputOutput supports interactions with OS for the plugin lib
+type osInputOutput interface {
+	// readOSArgs gets command line arguments passed to application
+	readOSArgs() []string
+	// printOut outputs given data to application standard output
+	printOut(data string)
+}
+
+// standardInputOutput delivers standard implementation for OS
+// interactions
+type standardInputOutput struct {
+}
+
+// tlsServerDefaultSetup provides default implementation for TLS setup routines
+type tlsServerDefaultSetup struct {
+}
+
+// tlsSetup holds TLS setup utility for plugin lib
+var tlsSetup tlsServerSetup = tlsServerDefaultSetup{}
+
+// libInputOutput holds utility used for OS interactions
+var libInputOutput osInputOutput = standardInputOutput{}
+
+// makeTLSConfig provides TLS configuraton template for plugins
+func (ts tlsServerDefaultSetup) makeTLSConfig() *tls.Config {
+	config := tls.Config{
+		ClientAuth:               tls.RequireAndVerifyClientCert,
+		PreferServerCipherSuites: true,
+		CipherSuites: []uint16{
+			tls.TLS_RSA_WITH_AES_128_GCM_SHA256,
+			tls.TLS_RSA_WITH_AES_256_GCM_SHA384,
+		},
 	}
-	rpc.RegisterStreamCollectorServer(server, proxy)
-	return startPlugin(server, m, &proxy.pluginProxy)
+	return &config
+}
+
+// readRootCAs delivers a standard source of root CAs from system
+func (ts tlsServerDefaultSetup) readRootCAs() (*x509.CertPool, error) {
+	return x509.SystemCertPool()
+}
+
+// readOSArgs implementation that returns application args passed by OS
+func (io standardInputOutput) readOSArgs() []string {
+	return os.Args
+}
+
+// printOut implementation that emits data into standard output
+func (io standardInputOutput) printOut(data string) {
+	fmt.Println(data)
+}
+
+// makeGRPCCredentials delivers credentials object suitable for setting up gRPC
+// server, with TLS optionally turned on.
+func makeGRPCCredentials(m *meta) (creds credentials.TransportCredentials, err error) {
+	var config *tls.Config
+	if !m.TLSEnabled {
+		config = &tls.Config{
+			InsecureSkipVerify: true,
+		}
+	} else {
+		cert, err := tls.LoadX509KeyPair(m.CertPath, m.KeyPath)
+		config = tlsSetup.makeTLSConfig()
```

Comment:
It seems that it doesn't make sense to make TLS Config if `tls.LoadX509KeyPair` returns error. Please consider it.